### PR TITLE
[promptflow] Add raise_on_error for stream run API

### PR DESCRIPTION
**Pull request by zhengfeiwang**
# Description

Add parameter `raise_on_error` for `pf.stream`, by default we raise for failed run.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py [hunk 1]
```diff
@@ -138,13 +140,17 @@ def stream(self, name: Union[str, Run]) -> Run:
...
             self._print_run_summary(run)
-            # print error message when run is failed
-            if run.status == RunStatus.FAILED:
-                error_message = local_storage.load_exception()["message"]
-                print_red_error(error_message)
         except KeyboardInterrupt:
             error_message = "The output streaming for the run was interrupted, but the run is still executing."
             print(error_message)
+
+        if run.status == RunStatus.FAILED:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_run_operations.py:147]: Canceled should also be checked?

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py [hunk 2]
```diff
@@ -111,11 +111,13 @@ def _print_run_summary(self, run: Run) -> None:
         )
 
     @monitor_operation(activity_name="pf.runs.stream", activity_type=ActivityType.PUBLICAPI)
-    def stream(self, name: Union[str, Run]) -> Run:
+    def stream(self, name: Union[str, Run], raise_on_error: bool = True) -> Run:
         """Stream run logs to the console.
 
         :param name: Name of the run, or run object.
         :type name: Union[str, ~promptflow.sdk.entities.Run]
+        :param raise_on_error: Raises an exception if a run fails.
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_run_operations.py:119]: add change for pfazure?

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py (hunk 1 above)
Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/operations/_run_operations.py]: updated on handling canceled run

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py (hunk 2 above)
Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/operations/_run_operations.py]: added

**Review by wangchao1230 — APPROVED** [review]

File: src/promptflow/tests/sdk_cli_azure_test/e2etests/test_run_operations.py [hunk 3]
```diff
@@ -91,21 +91,6 @@ def test_basic_evaluation(self, pf: PFClient, runtime: str, randstr: Callable[[s
         assert isinstance(eval_run, Run)
         pf.runs.stream(run=eval_run.name)
 
-        # evaluation run without data
```
Comment (0mza987) [inline src/promptflow/tests/sdk_cli_azure_test/e2etests/test_run_operations.py:94]: why delete this?

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py [hunk 4]
```diff
@@ -111,11 +111,13 @@ def _print_run_summary(self, run: Run) -> None:
         )
 
     @monitor_operation(activity_name="pf.runs.stream", activity_type=ActivityType.PUBLICAPI)
-    def stream(self, name: Union[str, Run]) -> Run:
+    def stream(self, name: Union[str, Run], raise_on_error: bool = True) -> Run:
```
Comment (0mza987) [inline src/promptflow/promptflow/_sdk/operations/_run_operations.py:114]: What's the context that we need to change this behavior? And shall we change azure stream behavior to make alignment?

Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/operations/_run_operations.py:114]: context: in local, if a run is failed or canceled, the logs might be empty, then `pf.stream` will show nothing; therefore, we add `raise_on_error` to raise exception in such case.

and I think I have synced the changes to azure stream?

File: src/promptflow/tests/sdk_cli_azure_test/e2etests/test_run_operations.py (hunk 3 above)
Comment (zhengfeiwang) [reply src/promptflow/tests/sdk_cli_azure_test/e2etests/test_run_operations.py:94]: as offline synced, I will update this test in another PR (will before this PR)

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py [hunk 5]
```diff
@@ -138,13 +140,20 @@ def stream(self, name: Union[str, Run]) -> Run:
...
+
+        if run.status == RunStatus.FAILED or run.status == RunStatus.CANCELED:
+            if run.status == RunStatus.FAILED:
+                error_message = local_storage.load_exception().get("message", "Run fails with unknown error.")
+            else:
+                error_message = "Run is canceled."
+            if raise_on_error:
+                raise InvalidRunStatusError(error_message)
+            else:
+                print_red_error(error_message)
```
Comment (D-W-) [inline src/promptflow/promptflow/_sdk/operations/_run_operations.py:155]: What's current behavior when pf run create with stream == True? Will we print duplicated error message?

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py [hunk 6]
```diff
@@ -138,13 +140,20 @@ def stream(self, name: Union[str, Run]) -> Run:
...
             self._print_run_summary(run)
-            # print error message when run is failed
-            if run.status == RunStatus.FAILED:
-                error_message = local_storage.load_exception()["message"]
-                print_red_error(error_message)
         except KeyboardInterrupt:
             error_message = "The output streaming for the run was interrupted, but the run is still executing."
             print(error_message)
+
+        if run.status == RunStatus.FAILED or run.status == RunStatus.CANCELED:
```
Comment (D-W-) [inline src/promptflow/promptflow/_sdk/operations/_run_operations.py:147]: Nit: can we share the logic between pfazure & pf?

**Review by 0mza987 — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py (hunk 5 above)
Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/operations/_run_operations.py:155]: I don't think there will be duplicate error message, and if the run failed, I think you will get an exception instead.

File: src/promptflow/promptflow/_sdk/operations/_run_operations.py (hunk 6 above)
Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/operations/_run_operations.py:147]: there are different ways to get error message for pf and pfazure, so I prefer not share them for now.

**Pull request merged**